src/load_csv.py

```diff
@@ -1,74 +1,119 @@
+from typing import List
 import hashlib
-
-import pandas as pd
+import csv
 from sqlalchemy import text
 
+from langchain_community.document_loaders.csv_loader import CSVLoader
+from langchain_core.documents import Document
+
 from dialog_lib.embeddings.generate import generate_embeddings
 from dialog.llm.embeddings import EMBEDDINGS_LLM
 from dialog_lib.db.models import CompanyContent
 from dialog.db import get_session
+from dialog.settings import Settings
+
+import logging
+
+logging.basicConfig(
+    level=Settings().LOGGING_LEVEL,
+    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
+)
+
+logger = logging.getLogger("make_embeddings")
 
 session = next(get_session())
+NECESSARY_COLS = ["category", "subcategory", "question", "content"]
+
+
+def _get_csv_cols(path: str) -> List[str]:
+    """Gets the csv columns from a csv file"""
+    with open(path) as f:
+        reader = csv.DictReader(f)
+        return reader.fieldnames
+
+def retrieve_docs_from_vectordb() -> List[Document]:
+    """Retrieve all documents from the vector store."""
+    company_contents: List[CompanyContent] = session.query(CompanyContent).all()
+    return [
+        Document(
+            page_content=content.content,
+            metadata={
+                "category": content.category,
+                "subcategory": content.subcategory,
+                "question": content.question,
+            },
+        )
+        for content in company_contents
+    ]
+
+def documents_to_company_content(doc: Document, embedding: float) -> CompanyContent:
+    """Transform each langchain's Document and its embedding to a CompanyContent object."""
+    return CompanyContent(
+        category=doc.metadata.get("category"),
+        subcategory=doc.metadata.get("subcategory"),
+        question=doc.metadata.get("question"),
+        content=doc.page_content,
+        embedding=embedding,
+        dataset=doc.metadata.get("dataset"),
+        link=doc.metadata.get("link"),
+    )
+
+
+def get_document_pk(doc: Document) -> str:
+    return (
+        doc.metadata["category"]
+        + doc.metadata["subcategory"]
+        + doc.metadata["question"]
+    )
+
 
 def load_csv_and_generate_embeddings(path, cleardb=False, embed_columns=("content",)):
-    df = pd.read_csv(path)
-    necessary_cols = ["category", "subcategory", "question", "content"]
-    for col in necessary_cols:
-        if col not in df.columns:
-            raise Exception(f"Column {col} not found in {path}")
+    metadata_columns = [col for col in _get_csv_cols(path) if col not in embed_columns]
 
-    if "dataset" in df.columns:
-        necessary_cols.append("dataset")
+    loader = CSVLoader(path, metadata_columns=metadata_columns)
+    docs: List[Document] = loader.load()
 
-    df = df[necessary_cols]
+    logger.info("Metadata columns: %s", metadata_columns)
+    logger.info("Embedding columns: %s", embed_columns)
+    logger.info("Glimpse over the first doc: %s", docs[0].page_content[:100])
```

Comment:
Do we need to make a print of the first document? Does it help us debugging?

Reply:
from my usual cases it really helps debbuging, because you want to check how the document column (or combination of them) are entering into the embeddings. So this is a suggestion for this debbuging.

Reply:
![embeddings](https://github.com/talkdai/dialog/assets/27077298/9dad846d-db5f-4a9e-9ac2-ff0262bf6ca2)

Reply:
an example:
```(dialog) ➜  dialog git:(update-load-csv-for-langchain-csv-loader) ✗ docker compose up dialog
[+] Running 2/0
 ✔ Container dialog-db-1      Running                                                                                                                  0.0s 
 ✔ Container dialog-dialog-1  Created                                                                                                                  0.0s 
Attaching to dialog-1
dialog-1  | INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
dialog-1  | INFO  [alembic.runtime.migration] Will assume transactional DDL.
dialog-1  | 2024-06-03 01:27:17,649 - make_embeddings - INFO - Metadata columns: ['category', 'subcategory', 'question', 'answer', 'actions']
dialog-1  | 2024-06-03 01:27:17,650 - make_embeddings - INFO - Embedding columns: ['content']
dialog-1  | 2024-06-03 01:27:17,650 - make_embeddings - INFO - Glimpse over the first doc: content: Question: Gratuidades e Meia Tarifa, desconto para idosos, desconto para estudantes, ID jov
dialog-1  | 2024-06-03 01:27:17,674 - root - INFO - Existing docs: 0
dialog-1  | 2024-06-03 01:27:17,674 - root - INFO - Generating embeddings for new questions...
dialog-1  | 2024-06-03 01:27:17,674 - root - INFO - New questions: 72
dialog-1  | 2024-06-03 01:27:17,674 - root - INFO - embed_columns: ['content']
dialog-1  | 2024-06-03 01:27:19,128 - httpx - INFO - HTTP Request: POST https://api.openai.com/v1/embeddings "HTTP/1.1 200 OK"
```